perf-event/src/lib.rs

```diff
@@ -813,6 +952,186 @@ impl IntoRawFd for Counter {
     }
 }
 
+impl Sampler {
+    /// Convert this sampler back into a counter.
+    ///
+    /// This is a one-way conversion, once you have the counter this crate
+    /// provides no way to open the kernel ring buffer again.
+    pub fn into_inner(self) -> Counter {
+        self.counter
+    }
+
+    /// Read the next record from the ring buffer.
+    ///
+    /// This method does not block. If you want blocking behaviour, use
+    /// [`next_blocking`] instead.
+    ///
+    /// It is possible to get readiness notifications for when events are
+    /// present in the ring buffer (e.g. for async code). See the documentation
+    /// on the [`perf_event_open`][man] manpage for details on how to do this.
+    ///
+    /// [`next_blocking`]: Self::next_blocking
+    /// [man]: https://man7.org/linux/man-pages/man2/perf_event_open.2.html
+    pub fn next(&mut self) -> Option<samples::Record> {
+        use crate::samples::{ParseBuf, ParseConfig, Record};
+        use bytes::Buf;
+        use memoffset::raw_field;
+        use std::{mem, ptr, slice};
+        use sys::bindings::{perf_event_header, perf_event_mmap_page};
+
+        let page = self.page();
+
+        // SAFETY:
+        // - page points to a valid instance of perf_event_mmap_page.
+        // - data_tail is only written by the user side so it is safe to do a
+        //   non-atomic read here.
+        let tail = unsafe { ptr::read(raw_field!(page, perf_event_mmap_page, data_tail)) };
```

Comment:
How is this different from `unsafe { *page }.data_tail`?

Reply:
I think that ends up either creating a reference (so maybe UB) or copying the whole struct and then grabbing a field (probably also UB since it's being modified concurrently). I'm a little out of my depth here on what's UB or not so I've decided to play it safe.

Reply:
Also, I'm not sure that using a non-atomic read here satisfies the [contract for walking the ring buffer](https://github.com/torvalds/linux/blob/af7a056891899fd3942afec79fb219f58271e319/tools/include/linux/ring_buffer.h#L7-L49) in the kernel sources.

Reply:
The contract says nothing about user-space reading `data_tail` since the kernel never writes to it.

FWIW, the perf event reader in perfetto [makes the same choice](https://github.com/google/perfetto/blob/master/src/profiling/perf/event_reader.cc#L151)

Reply:
Okay. Perfetto is widely used, so I'm inclined to trust that. I'm still amazed by this "access field of struct" pixie dust, but we can keep it for now.